firmware/src/PeO.c

```diff
@@ -0,0 +1,102 @@
+#include "PeO.h"
+#ifdef MACHINE_ON
+    #include "machine.h"
+#endif
+#include "usart.h"
+
+volatile float max_power;
+volatile float max_power_duty_cycle;
+static uint8_t done;
+static uint8_t callSweep;
+volatile uint8_t count;
+
+void perturb_and_observe(void){
+    static uint8_t d_step = PWM_D_STEP;
+
+    // Computes power input
+    control.pi[0] = control.v_panel[0] * control.i_panel[0];
+
+    //Derivate power
+    float dpi = (control.pi[0]) -(control.pi[1]);
+    float dvi = (control.v_panel[0]) -(control.v_panel[1]);
+    float dii = (control.i_panel[0]) - (control.i_panel[1]);
+    float coef_di = (dii/dvi);
+    float coef_def = ((-control.i_panel[0]/control.v_panel[0]));
+
+ 
+    if(!callSweep){
+        if(dvi != 0){
+            if(coef_di != coef_def){
```

Comment:
This conditionals make sense? because is really hard to two floats be **exactly** equal, then this condition almost always will be true

Reply:
if not sure test adding a log on the else condition to see if is triggered in some condition